specs/src/recovery.md

```diff
@@ -0,0 +1,525 @@
+data.
+
+The problem arises when applications have less time to distribute data than they
+can avoid or remove an attacker. Recovery is such as case. If an attacker wants
+to disrupt a network, it only needs to slow the distribution of data
+sufficiently beyond the `ProposalTimeout`. While the network will eventually
+distribute that data, the throughput will be dramatically decreased until
+connections to malicious peers are removed.
+
+> Side note: its easy to confuse a congested peer as a malicious one. This makes
+> peer scoring incredibly difficult for pull based systems. If timeouts for
+> scoring or addional requests are too agressive, feedback loops emerge.
+> Congested nodes can hit timeouts, resulting in requesting more duplicate data,
+> which causes more congestion.
+
+**Conclusions:**
+
+- Time sensitive applications require additional mechanisms in order to use pull
+  based gossip.
+
+#### Broadcast Trees
+
+Using the most naive design of broadcast trees, where the broadcaster
+distributes equal amounts of block data to each of its peers, a single faulty
+peer will break the first constraint of recovery. Even if parity data is added,
+determined attackers can simply spin up a large majority of nodes in the network
+and not propagate data.
+
+**Conclusions:**
+
+1) The usage of broadcast trees requires mechanisms that allow for faulty nodes
+   to exist.
+2) Validators must be able to differentiate between themselves and
+   non-validators. Without this, attackers can also increase the number of nodes
+   in the network and poison the tree.
+
+### Defending Against Attacks
+
+In order to prevent attackers from sending `Have` messages to all peers and simply not responding, we use the `PerPeerConcurrentRequestCap` discussed above.
+
+```go
+// the largest number of validators to compose 1/3 of the voting power. If we
+// assume even voting power percentage, then that is simply 1/3 of the number of
+// validators.
+faultyValidators := max(1, validatorSet.Size() / 3)
+
+// the faultyPartCount is determined by amount of parity data added
+faultyPartCount = math.Abs(parts.Total() - parityParts.Total())
+
+concurrentPartLimit = faultyPartCount / faultyValCount
+```
+
+The above attacks all rely on an attacker's ability to add many malicious nodes
+to the network. Our strategy towards defending against attacks will likely
+involve mechanisms that prevent this from occurring in the first place,
+alongside ones that mitigate damage when it does.
+
+#### Erasure Encoding Recovery Data
+
+Distributing erasure encoded data alongside the original data introduces
+interesting tradeoffs to the attacks above.
+
+```go
+var (
+    // originalData is the block data without erasure encoding
+    originalData = []byte{...}
+    // totalData include parity data
+    totalData = []byte{...}
+)
+
+// e is the erasure encoding factor that specifies the ratio of parity data
+e := len(totalData) / len(originalData) 
```

Comment:
Is there some recommended value of e?

Reply:
Assuming this is the same erasure coding we use currently? In that case `e` would be 4 (as `len(totalData)==4*len(originalData)`